r-lib-error-1.0.md

```diff
@@ -0,0 +1,60 @@
+
+# Specification of r-lib errors, version 1.0
+
+Note: this is an experimental draft.
+
+Note: version 1.0 corresponds to the `rlang_error` class, as
+implemented in rlang 0.4.0.
+
+## Classes
+
+An r-lib error object always has at least classes `rlib_error`, `error`
+and `condition`. It may have other classes.
+
+## Members
+
+An `rlib_error` object has the following members:
+
+* `message`: `character(1)`, the error message.
```

Comment:
Do we actually want to allow that? A bunch of r-lib code just uses `err$message` instead of the official `conditionMessage()` API.

Reply:
I actually think we should consider not _requiring_ a `message` member at all, just use `conditionMessage()`.

Reply:
I am up for that, but it will potentially break a lot of current r-lib code and docs that uses `$message`. E.g.
https://github.com/r-lib/rlang/blob/bc1cb38c3083e912042a76d0fd3fd7f95c69982d/R/cnd-abort.R#L91
https://github.com/r-lib/rlang/blob/4593159964e264515ac56f8f1f78d000caa38a15/R/cnd-entrace.R#L54
or even `testthat::expect_error()`.

Reply:
We're moving towards `conditionMessage()` (or `rlang::cnd_message()` for the bullets format). `testthat::expect_error()` is fixed in dev testthat already.

Thanks for pointing these in rlang, I'll fix before sending 0.4.1.

Reply:
That's great. There'll be a lot more to fix, though.

Reply:
If it helps here are the counts of using `$message` in r-lib packages, at least some of which are false positives. The full results are at https://gist.github.com/jimhester/9ba97fb3a090c7c182db6effbfceca50

```
testthat 20
rlang 9
async 7
callr 7
pkgcache 7
liteq 6
roxygen2 5
gargle 4
lifecycle 4
remotes 4
cli 3
credentials 3
gmailr 3
processx 3
progress 3
styler 3
vctrs 3
vdiffr 3
cliapp 2
covr 2
devtools 2
evaluate 2
httr 2
pkgdepends 2
conflicted 1
desc 1
gh 1
installgithub.app 1
later 1
oldie 1
pkgbuild 1
pkgdown 1
pkgload 1
zeallot 1
```